Source/Core/Common/Intrinsics.h

```diff
@@ -4,28 +4,90 @@
 
 #pragma once
 
-#ifdef _M_X86
+#if defined(_M_GENERIC)
 
-#ifdef _MSC_VER
+// We don't support intrinsics in generic builds.
+
+#elif defined(_M_X86)
+
+// Due to limitations in GCC, SSE intrinsics are only available when compiling with the
+// corresponding instruction set enabled. However, using the target attribute, we can compile
+// single functions with a different target instruction set, while still creating a generic build.
+//
+// Since this instruction set is enabled per-function, any callers should verify that the
+// instruction set is supported at runtime before calling it, and provide a fallback implementation
+// when not supported.
+//
+// SUPPORTS_FUNCTION_TARGET_<level> determines which intrinsics are available for use in any
+// function that specifies the instruction set via FUNCTION_TARGET_<level>.
+
+#if defined(_MSC_VER) || defined(__INTEL_COMPILER)
+
+// MSVC and ICC support intrinsics for any instruction set without any function attributes.
 #include <intrin.h>
-#else
+#define SUPPORTS_FUNCTION_TARGET_SSE42 1
+#define FUNCTION_TARGET_SSE42
+#define SUPPORTS_FUNCTION_TARGET_SSE41 1
+#define FUNCTION_TARGET_SSE41
+#define SUPPORTS_FUNCTION_TARGET_SSSE3 1
+#define FUNCTION_TARGET_SSSE3
+#define SUPPORTS_FUNCTION_TARGET_SSE3 1
+#define FUNCTION_TARGET_SSE3
+
+#elif defined(__GNUC__) || defined(__clang__)
+
+// When building with -march=native, or enabling the instruction sets in the compile flags, permit
+// usage of the instrinsics without any function attributes.
 #include <x86intrin.h>
+#ifdef __SSE4_2__
+#define SUPPORTS_FUNCTION_TARGET_SSE42 1
+#define FUNCTION_TARGET_SSE42
+#endif
+#ifdef __SSE4_1__
+#define SUPPORTS_FUNCTION_TARGET_SSE41 1
+#define FUNCTION_TARGET_SSE41
+#endif
+#ifdef __SSSE3__
+#define SUPPORTS_FUNCTION_TARGET_SSSE3 1
+#define FUNCTION_TARGET_SSSE3
+#endif
+#ifdef __SSE3__
+#define SUPPORTS_FUNCTION_TARGET_SSE3 1
+#define FUNCTION_TARGET_SSE3
 #endif
 
-#if defined _M_GENERIC
-#define _M_SSE 0
-#elif _MSC_VER || __INTEL_COMPILER
-#define _M_SSE 0x402
-#elif defined __GNUC__
-#if defined __SSE4_2__
-#define _M_SSE 0x402
-#elif defined __SSE4_1__
-#define _M_SSE 0x401
-#elif defined __SSSE3__
-#define _M_SSE 0x301
-#elif defined __SSE3__
-#define _M_SSE 0x300
+// Support for function target attributes requires GCC 4.9+ or clang 3.7.0+.
```

Comment:
AFAIK we already require gcc 4.9. Through still without cmake check.
And after looking at the fallback code, I think it's fine to also require 4.9 here.